### PR TITLE
What a selection operator could look like

### LINKED ISSUE #270: Provide grb::select( [out] Matrix, [in] Matrix, grb::operator(r,c,v) -> bool )
The `select` primitive(s) is required for multiple incoming algorithms. 
For instance to threshold under/over a specified value, or in `tril` in order to filter depending on the coordinates.

The idea would be to generalise the incoming `tril`, `triu` and `diag` primitives.
- `grb::tril( L, A )` -> `grb::select( L, A, grb::boolean_operators::is_lower_part<RIT, CIT>() )`
- `grb::triu( U, A )` -> `grb::select( U, A, grb::boolean_operators::is_upper_part<RIT, CIT>() )`
- `grb::diag( D, A )` -> `grb::select( D, A, grb::boolean_operators::is_diag<RIT, CIT>() )`
- `grb::diag( A )` -> `grb::select( A, grb::boolean_operators::is_diag<RIT, CIT>() )`

The **diag** boolean_operator would then have the following `apply` function:
```c++
template< 
  typename RIT, typename CIT, typename T,
  typename BoolOrEquivalent = bool
>
BoolOrEquivalent apply( RIT i, CIT j, T value ) {
  return static_cast<BoolOrEquivalent>( i == j );
}
```

The naming is to be discussed, of course.

### PR DESCRIPTION
**Pull request by byjtew**
Closes #270 

- [x] Wrapping the selection operators in their own namespace to avoid confusion: `grb::operators::select::`
- [x] Implement selectLambda
- [x] Distributed backends implementation

- In-place variants (should come in a separate PR)
- Masked variants (should come in a separate PR)
- Pattern matrices support (should come in a separate PR)

### CODE REVIEW

**Comment by byjtew** [comment]
Next steps:
- [ ] Wrapping the selection operators in their own namespace to avoid confusion
- [ ] Implement selectLambda
- [ ] In-place variants (may be in a separate PR)
- [ ] Masked variants (may be in a separate PR)

**Comment by anyzelman** [comment]
The current state does not have bsp1d nor hybrid implementations of select, that's correct right? I think this is a super useful enhancement, but before merge it should have those distributed-memory implementations also

**Comment by byjtew** [comment]
> The current state does not have bsp1d nor hybrid implementations of select, that's correct right? I think this is a super useful enhancement, but before merge it should have those distributed-memory implementations also

Absolutely, I also wanted to switch the namespace from `operators::` to `select_operators::` or `operators::select::` since they have different signatures than usual `operators::`

I'll ping you once it will be implemented

**Comment by anyzelman** [comment]
> > The current state does not have bsp1d nor hybrid implementations of select, that's correct right? I think this is a super useful enhancement, but before merge it should have those distributed-memory implementations also
> 
> Absolutely, I also wanted to switch the namespace from `operators::` to `select_operators::` or `operators::select::` since they have different signatures than usual `operators::`
> 
> I'll ping you once it will be implemented

maybe just `grb::select`?

**Comment by byjtew** [comment]
> maybe just `grb::select`?

But then it could be either the primitive or the operator's namespace, what about `grb::select::operators` ?

**Comment by anyzelman** [comment]
> > maybe just `grb::select`?
> 
> But then it could be either the primitive or the operator's namespace, what about `grb::select::operators` ?

Oh right.. It could still work by having select as a class with the "primitive call" via `operator()` and the pre-defined lambdas as public static members, but that's perhaps a bit hacky. Maybe just think of something and we can look at it in more detail later, once it's more or less ready?

**Comment by anyzelman** [comment]
Also doc/spec should be added to base/blas3.hpp

File: include/graphblas/base/blas3.hpp
Comment (anyzelman) [inline include/graphblas/base/blas3.hpp:550]: isn't it nicer to typedef `SelectionOperator` as an `std::function` with a specific signature, which would then no longer require `selectLambda` as a separate primitive? (Different from `grb::algorithms`, we want to have an absolute minimum number of core primitives.)

File: include/graphblas/base/blas3.hpp
Comment (anyzelman) [inline include/graphblas/base/blas3.hpp:538]: Doc is still missing?

File: include/graphblas/base/blas3.hpp
Comment (anyzelman) [inline include/graphblas/base/blas3.hpp]: tab instead of a space before `=` (elsewhere also)

File: include/graphblas/base/blas3.hpp
Comment (anyzelman) [reply include/graphblas/base/blas3.hpp:550]: I now remember a concern about the performance of lambda functions and std::functions in particular.

There is another alternative, which is to specify that the `SelectionOperator` is a functor, i.e., a class of the following form:

```
template< typename RIT, typename CIT, typename D >
class MySelectionOperator {

    public:

        static bool operator()( const RIT i, const CIT j, const D v ) {
               // some logic based on i, j, v returning a bool
        }

};
```

I would have to check, but I suspect that the use of this (i.e., in the select function, there would be calls like `if( SelectionOperator( i, j, v ) ) { ... }`) would work even if SelectionOperator were an std::function or a lambda function, isn't it?

File: include/graphblas/base/internalops.hpp
Comment (anyzelman) [inline include/graphblas/base/internalops.hpp]: If following the above suggestion, then these base classes are not required

File: include/graphblas/base/internalops.hpp
Comment (anyzelman) [inline include/graphblas/base/internalops.hpp]: If following the above suggestion, then instead of `apply(`, these should be `static bool operator()(`

File: include/graphblas/ops.hpp
Comment (anyzelman) [inline include/graphblas/ops.hpp]: I suggest `namespace selectors` (i.e., put the functors in `grb::selectors`)

File: include/graphblas/ops.hpp
Comment (anyzelman) [inline include/graphblas/ops.hpp]: If following the above suggestion, then a direct functor implementation in `ops.hpp` that does not require this inheritance + reflection becomes possible.

The reason why ops.hpp is more complicated is because 1) it automatically derives foldl/foldr and array-based operations when no explicit ones are given, 2) because there is an algebraic type traits system, and 3) because the operators need to be composed into richer algebraic constructs. I don't think we have any of such a view or requirement for selectors right, so we can pick a simpler solution instead?

File: include/graphblas/reference/blas3.hpp
Comment (anyzelman) [inline include/graphblas/reference/blas3.hpp]: `(void)` before `nzc++`

File: include/graphblas/reference/blas3.hpp
Comment (anyzelman) [inline include/graphblas/reference/blas3.hpp]: `firstprivate` copies the given values, which in this case seem unnecessary copies. The default share-everything would be more performant?

File: include/graphblas/reference/blas3.hpp
Comment (anyzelman) [inline include/graphblas/reference/blas3.hpp]: I don't think the OpenMP standard specifies the performance semantics of reduction -- it is likely to be implementation-dependent. Therefore an OpenMP run-time implementation may dynamically allocate memory to facilitate the reduction. Instead relying on the internal ALP/GraphBLAS buffer and explicitly doing the reduction prevents this possibility and ensures the implementation adheres to the performance semantics (which, although presently not specified, is likely to include the promise that no system calls such as malloc or free are made).

File: include/graphblas/reference/blas3.hpp
Comment (anyzelman) [inline include/graphblas/reference/blas3.hpp]: note `firstprivate` again (possibly elsewhere as well)

In this case also `in_coordinates` seem unused. In particular the use of `default(none)` in my experience always ends up annoying to maintain because code changes more often than one might think affects the shared and private clauses-- which might be what happened here (if an earlier version did use `in_coordinates`)

File: include/graphblas/reference/blas3.hpp
Comment (anyzelman) [inline include/graphblas/reference/blas3.hpp]: `(void)` here (and elsewhere possibly also)

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: line overflow?

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: use `!` instead of `not`

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: spaces after (before) < (>)

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: spaces after `(` and before `)`

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: here and perhaps elsewhere too

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: line overflow?

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp:652]: missing empty line at the bottom?

File: include/graphblas/bsp1d/blas3.hpp
Comment (anyzelman) [inline include/graphblas/bsp1d/blas3.hpp]: I don't think this is a correct implementation -- BSP1D employs a block-cyclic distribution, so local coordinates (i,j) are not simply an offset away from the global coordinates. Other primitives (like mxv) resolve this by passing a lambda that describes how to translate local to global coordinates. That lambda could replace the offset that you currently have implemented. Also, the test should likely be expanded to run into this error (apparently at present it does not right?)

Comment (anyzelman) [reply include/graphblas/bsp1d/blas3.hpp]: Passing in a generic distribution / how to get from local to global has the added benefit that it will work with any other distributed-memory backend (which may make other distribution choices)

**Review by anyzelman — CHANGES_REQUESTED** [review]
Thanks for this useful primitive!

I think this one can be simplified as in the comments. The suggested simplifications should benefit both usage as well asimplementation.

File: include/graphblas/base/blas3.hpp
Comment (byjtew) [reply include/graphblas/base/blas3.hpp:550]: Indeed and I wanted to do that at first, but I copied the logic of the current operators in case someone absolutely needed to write the word *apply*

File: include/graphblas/base/internalops.hpp
Comment (byjtew) [reply include/graphblas/base/internalops.hpp]: Indeed

File: include/graphblas/base/blas3.hpp
Comment (byjtew) [reply include/graphblas/base/blas3.hpp:550]: Done.

File: include/graphblas/base/internalops.hpp
Comment (byjtew) [reply include/graphblas/base/internalops.hpp]: For the void specialisation they are

File: include/graphblas/ops.hpp
Comment (byjtew) [reply include/graphblas/ops.hpp]: Type handling is easier this way, otherwise you need to specify it explicitly

File: include/graphblas/base/blas3.hpp
Comment (anyzelman) [inline include/graphblas/base/blas3.hpp]: I think the latter variant was never implemented right? (I'm deleting it during review, please let me know if that was premature @byjtew )

Comment (byjtew) [reply include/graphblas/base/blas3.hpp]: Absolutely

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: @byjtew to check, this comment is outdated right?

File: tests/unit/selectMatrix.cpp
Comment (anyzelman) [inline tests/unit/selectMatrix.cpp]: And this comment too?

**Comment by anyzelman** [comment]
Note: during review, I was trying out lambda-based selection. I'm working on debugging what happens when modifying the unit tests to test that functionality, while also working on simplifying the interface to cope with both pre-defined selection ops as well as arbitrary lambdas. It's getting late though, so will continue this tomorrow.

File: tests/unit/selectMatrix.cpp
Comment (byjtew) [reply tests/unit/selectMatrix.cpp]: Yes

File: tests/unit/selectMatrix.cpp
Comment (byjtew) [reply tests/unit/selectMatrix.cpp]: Yes

**Comment by anyzelman** [comment]
This one is ready to merge (after conflict resolution)

**Comment by anyzelman** [comment]
> This one is ready to merge (after conflict resolution)

Correction: unit tests still detect a bug in BSP1D, fixing it is TODO

**Comment by anyzelman** [comment]
Fixed

**Review by anyzelman — APPROVED** [review]

**Comment by anyzelman** [comment]
Some final tests (with LPF) are running, as is the CI. Will merge if all OK.

Concept release notes:

This MR introduces the matrix selection feature -- given a operator that takes matrix nonzero coordinates and a value as arguments and produces a boolean, elements from an input matrix are selected to be copied in some output matrix. This primitive operates in an out-of-place fashion. In-place semantics may be achieved via a temporary (pattern) matrix and fold, and thus are currently not planned to be added. Standard matrix selection operators such as (strictly) lower triangular etc. are provided, including type traits for them. Non-standard selection operators provided by users are simple to add as the new `grb::select` primitive accepts arbitrary lambda functions as a selection operator as well.

This MR also fixes a bug for the BSP1D and hybrid eWiseApply on matrices with the resize mode, which could return `SUCCESS` even if it was not successful. The MR includes unit tests for matrix distributions and the new select operator, as well as code style fixes in related and unrelated files.

**Pull request merged**
